### PR TITLE
Add proper readme explaining concepts and usage

### LINKED ISSUE #18: Where are the docs?
Can’t seem to find usage docs for this component. Where should I be looking?

### PR DESCRIPTION
**Pull request by bigs**
Will eventually close #18 #9 #5.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,16 +12,274 @@
...
+
+The testlab CLI depends on the presence of the standard environment variables to
+connect out to your Nomad and Consul clusters. If any are ommitted, the
+defaults, as defined by Hashicorp, will be applied. The defaults are typically
+usable in development.
+
+- [Nomad environment variables](https://github.com/hashicorp/nomad/blob/v0.9.3/api/api.go#L217)
+- [Consul environment variables](https://github.com/hashicorp/consul/blob/v1.5.2/api/api.go#L24)
+
+Furthermore, users can optionall provide a path in the environment variable
```
Comment (raulk) [inline README.md]:
```suggestion
Furthermore, users can optionally provide a path in the environment variable
```

File: README.md [hunk 2]
```diff
@@ -12,16 +12,274 @@
...
+`TESTLAB_ROOT` to define where the testlab metadata will be stored. This
+defaults to `/tmp/testlab`. **NOTE**: In order to have multiple testlab
+topologies in flight at the same time, one must define different `TESTLAB_ROOT`s
+for each topology.
+
+The testlab CLI has but two commands:
```
Comment (raulk) [inline README.md]:
```suggestion
The testlab CLI has two commands:
```

File: README.md [hunk 3]
```diff
@@ -12,16 +12,274 @@
...
+Defines how many of this type of peer should be launched in the cluster.
+
+##### `Options: object`
+
+An optional object as defined by the specific
+[node implementation](#node-implementations).
+
+##### `Dependencies: list`
+
+A list of **`Name`**s of deployments that must be scheduled before this one.
```
Comment (raulk) [inline README.md]: The `Name` formatting doesn't work :-(

File: README.md [hunk 4]
```diff
@@ -12,16 +12,274 @@
...
+before generic peers that might want to bootstrap on them, or ensuring a
+deployment of peers is launched before. The scenario that drives them is
+scheduled. Cycles are not permitted.
+
+### Scenario Runners
+
+Scenario runners are the beating heart of testlab's simulation capabilities.
+It is their responsibility to drive the various deployments to create activity
+within the network. While it's not entirely necessary to use the `scenario` node
+to deploy drivers, it can be quite useful, especially in larger clusters.
```
Comment (raulk) [inline README.md]: The term driver is a little ambiguous here... to deploy _test cases_, maybe?

File: README.md [hunk 5]
```diff
@@ -12,16 +12,274 @@
...
+
+The scenario runner API is described by its
+[node implementation](testlab/node/scenario/scenario.go) **and is, at present, a
+work in progress**. Pull requests welcome!
```
Comment (raulk) [inline README.md]: What kinds of things/features would we like pull requests for?

File: README.md [hunk 6]
```diff
@@ -12,16 +12,274 @@
...
+
+Scenario runners can expect a few environment variables to be present, to aid
+them in connecting to the peers they wish to control. These variables are mostly
+tailored towards helping them interact with Consul, to discover information
+about the peers they've been assigned to.
+
+- `DAEMON_CLIENTS` (int): The number of ports this scenario runner has been
```
Comment (raulk) [inline README.md]: ports? in which sense?

File: README.md [hunk 7]
```diff
@@ -12,16 +12,274 @@
 
 Check out the [ROADMAP](ROADMAP.md) to see what's coming.
 
+## Table of Contents
+
+- [Dependencies](#dependencies)
```
Comment (raulk) [inline README.md:21]: since the readme is long, maybe generate a more detailed ToC with https://github.com/thlorenz/doctoc?

**Review by raulk — COMMENTED** [review]
This is shaping up good! 🙌

File: README.md (hunk 7 above)
Comment (bigs) [reply README.md:21]: was looking for something like this!

File: README.md (hunk 5 above)
Comment (bigs) [reply README.md]: i'll add a comment at the bottom! i think more p2p app support

**Review by bigs — COMMENTED** [review]
ready for a final review @raulk @lanzafame

File: README.md [hunk 8]
```diff
@@ -12,23 +12,359 @@
...
+defaults, as defined by Hashicorp, will be applied. The defaults are typically
+usable in development.
+
+- [Nomad environment variables](https://github.com/hashicorp/nomad/blob/v0.9.3/api/api.go#L217)
+- [Consul environment variables](https://github.com/hashicorp/consul/blob/v1.5.2/api/api.go#L24)
+
+Furthermore, users can optionally provide a path in the environment variable
+`TESTLAB_ROOT` to define where the testlab metadata will be stored. This
+defaults to `/tmp/testlab`. **NOTE**: In order to have multiple testlab
+topologies in flight at the same time, one must define different `TESTLAB_ROOT`s
```
Comment (lanzafame) [inline README.md:150]: Is this a hard requirement? Not really an issue with this documentation PR but I will extract into a separate issue if it isn’t a hard requirement. If it is a hard requirement, could the README.md get a short explanation as to why?

File: README.md [hunk 9]
```diff
@@ -12,23 +12,359 @@
...
+for each topology.
+
+The testlab CLI has two commands:
+
+- `testlab start <json configuration>`
+  Parses, evaluates for correctness, and attempts to deploy a topology as
+  defined by the provided json configuration file.
```
Comment (lanzafame) [inline README.md]: And then attempt to run the `scenarios`?

File: README.md [hunk 10]
```diff
@@ -12,23 +12,359 @@
...
+- `testlab stop`
+  Stops the current running topology, identified by its `TESTLAB_ROOT`.
+
+### Deployment Configuration
+
+The entrypoint for most projects using the testlab will be their deployment
+configuration, a JSON document declaring the desired network configuration.
+It's broken into the following top level sections:
+
+#### `Name: string`
```
Comment (lanzafame) [inline README.md:171]: Having the JSON fields and their types split over the sub-headings is rather confusing. Maybe have a non-populated “json schema” at the top to show the sections and then just use the fields names as the sub-headings.

File: README.md [hunk 11]
```diff
@@ -12,23 +12,359 @@
...
+    // allows nomad to determine which tasks should be scheduled when there is
+    // resource contention. If your nomad cluster has other tasks running on it,
+    // be sure to set this value accordingly. Otherwise, a default of 50 will be
+    // provided.
+    "Priority": int,
+}
+```
+
+#### `Deployments: list of objects`
+
```
Comment (lanzafame) [inline README.md:197]: Again, a snippet of the json object for easy parsing, also it allows you to show an example configuration.

File: README.md [hunk 12]
```diff
@@ -12,23 +12,359 @@
...
+	utils "github.com/libp2p/testlab/utils"
+)
+
+type Node interface {
+	Task(utils.NodeOptions) (*napi.Task, error)
+	PostDeploy(*capi.Client, utils.NodeOptions) error
+}
+```
+
+Given some `NodeOptions`, a wrapper over the `map[string]interface{}` type
```
Comment (lanzafame) [inline README.md]:
```suggestion
Given some `utils.NodeOptions`, a wrapper over the `map[string]interface{}` type
```

File: testlab/testlab.go [hunk 13]
```diff
@@ -12,6 +12,7 @@ var testLab *testlab.TestLab
 
 func main() {
 	app := cli.NewApp()
+	app.Version = "0.1-alpha"
```
Comment (lanzafame) [inline testlab/testlab.go]: semver?

**Review by lanzafame — CHANGES_REQUESTED** [review]

File: README.md (hunk 9 above)
Comment (bigs) [reply README.md]: technically the scenarios are part of the topology, but i'll mention that.

File: README.md (hunk 10 above)
Comment (bigs) [reply README.md:171]: yeah i will probably do a json schema in a subsequent revision. i wasn't sure what the best way to have it be thoroughly annotated in place was

**Comment by bigs** [comment]
@raulk or @lanzafame feel free to merge this once you've had a chance to look.

File: README.md (hunk 10 above)
Comment (lanzafame) [reply README.md:171]: The types have now propagated to the TOC, since you generate the TOC...

Comment (lanzafame) [reply README.md:171]: At this point just an example config would be better than a full json schema.

Comment (bigs) [reply README.md:171]: wow i really should have linked to my example config... will do.

File: README.md (hunk 8 above)
Comment (bigs) [reply README.md:150]: not a hard requirement, just a MVP quality. can easily be extended to support multiple simultaneous topologies

**Review by lanzafame — APPROVED** [review]

**Pull request merged**